### PR TITLE
[SPARK-14963][Yarn] Using recoveryPath if NM recovery is enabled

### PR DESCRIPTION
**Pull request by jerryshao**
## What changes were proposed in this pull request?

From Hadoop 2.5+, Yarn NM supports NM recovery which using recovery path for auxiliary services such as spark_shuffle, mapreduce_shuffle. So here change to use this path install of NM local dir if NM recovery is enabled.
## How was this patch tested?

Unit test + local test.

### CODE REVIEW

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java [hunk 1]
```diff
@@ -222,4 +219,43 @@ protected void serviceStop() {
...
+   * Set the recovery path for shuffle service recovery when NM is restarted. The method will be
+   * overrode and called when Hadoop version is 2.5+ and NM recovery is enabled, otherwise we
+   * have to manually call this to set our own recovery path.
+   */
+  public void setRecoveryPath(Path recoveryPath) {
+    _recoveryPath = recoveryPath;
+  }
+
+  /**
+   * Get the recovery path, this will override the default one to get the our own maintained
```
Comment (tgravescs) [inline common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java]: nit: "to get our" remove "the"

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java [hunk 2]
```diff
@@ -112,14 +118,15 @@ private boolean isAuthenticationEnabled() {
...
+    _conf = conf;
 
     // In case this NM was killed while there were running spark applications, we need to restore
     // lost state for the existing executors.  We look for an existing file in the NM's local dirs.
     // If we don't find one, then we choose a file to use to save the state next time.  Even if
     // an application was stopped while the NM was down, we expect yarn to call stopApplication()
     // when it comes back
     registeredExecutorFile =
-      findRegisteredExecutorFile(conf.getTrimmedStrings("yarn.nodemanager.local-dirs"));
+      new File(getRecoveryPath().toUri().getPath(), "registeredExecutors.ldb");
```
Comment (tgravescs) [inline common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java]: I know you didn't change this but could you make a constant for "registeredExecutors.ldb"

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java [hunk 3]
```diff
@@ -222,4 +219,43 @@ protected void serviceStop() {
...
+    String[] localDirs = _conf.getTrimmedStrings("yarn.nodemanager.local-dirs");
+    for (String dir : localDirs) {
+      File f = new File(new Path(dir).toUri().getPath(), "registeredExecutors.ldb");
+      if (f.exists()) {
+        if (_recoveryPath == null) {
+          // If NM recovery is not enabled, we should specify the recovery path using NM local
+          // dirs, which is compatible with the old code.
+          _recoveryPath = new Path(dir);
+        } else {
+          // If NM recovery is enabled and recovery file is existed in old NM local dirs, which
```
Comment (tgravescs) [inline common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java]: nit change "and recovery file is existed in old" to "the recovery file exists in the old"

File: yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceSuite.scala [hunk 4]
```diff
@@ -234,7 +236,25 @@ class YarnShuffleServiceSuite extends SparkFunSuite with Matchers with BeforeAnd
     s3.initializeApplication(app2Data)
     ShuffleTestAccessor.getExecutorInfo(app2Id, "exec-2", resolver3) should be (Some(shuffleInfo2))
     s3.stop()
-
   }
 
+  test("get correct recovery path") {
```
Comment (tgravescs) [inline yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceSuite.scala]: could you also add a test to make sure the move happens properly when upgrading.

**Comment by tgravescs** [comment]
few minor comments but mostly looks good.  Did you build against both hadoop 2.5+ and hadoop < 2.5?

Did you manually test the upgrade path?

**Comment by jerryshao** [comment]
Thanks @tgravescs for your comments, I will change the code and do a more comprehensive test accordingly.

**Comment by jerryshao** [comment]
@tgravescs , I tested locally using Hadoop 2.4 and 2.6 with different scenarios:

1. Only Hadoop 2.4
2. Hadoop 2.4 upgrade to 2.6 with NM recovery disabled.
3. Hadoop 2.4 upgrade to 2.6 with NM recovery enabled.
3. Hadoop 2.6 NM recovery disabled to enabled.

Looks fine in all these scenarios.

One missing part is do we need to take care of downgrade scenarios, like 2.6 to 2.4 or NM recovery enabled to disabled?

**Comment by jerryshao** [comment]
Jenkins, retest this please.

**Comment by tgravescs** [comment]
I'm not concerned with the downgrade case.  It just won't find the file if yarn isn't setting the recovery path any longer (it will create new one in localdir) , but I don't see that as a big issue because if someone is downgrading their cluster or turned off recovery they should kill everything that is running.

**Comment by tgravescs** [comment]
+1  Thanks @jerryshao

**Pull request closed without merging**